### PR TITLE
datetime-diff function for MongoDB

### PR DESCRIPTION
**Pull request by calherries**
Implements the datetime-diff function for MongoDB.

Epic: https://github.com/metabase/metabase/issues/26999
Original datetimeDiff PR: https://github.com/metabase/metabase/pull/25722

(note for review: I actually want to merge this into master, not mongo-remove-v3-support, but it depends on that PR merging)

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/27042)

### CODE REVIEW

**Comment by calherries** [comment]
This is waiting on https://github.com/metabase/metabase/pull/27110 and the issue in this comment https://github.com/metabase/metabase/pull/26983#issuecomment-1347141918 to be resolved

**Comment by calherries** [comment]
Note this PR depends on https://github.com/metabase/metabase/pull/27176, since `driver/database-supports?` will return false until this is fixed.

File: modules/drivers/mongo/src/metabase/driver/mongo.clj [hunk 1]
```diff
@@ -246,6 +246,11 @@
                    :semantic-version
                    (driver.u/semantic-version-gte [5]))))
 
+(defmethod driver/database-supports? [:mongo :datetime-diff] [_ _ db]
+  (boolean (some-> (:dbms_version db)
+                   :semantic-version
+                   (driver.u/semantic-version-gte [5]))))
```
Comment (metamben) [inline modules/drivers/mongo/src/metabase/driver/mongo.clj:252]: Nil punning has been removed on the v3 removal branch.

**Review by metamben — COMMENTED** [review]
Some test will be added, right?

**Comment by calherries** [comment]
> Some test will be added, right?

The tests are in the "Datetime diff tests" block here https://github.com/metabase/metabase/blob/2eb0f03a10efd5a4a77260a85e3fa0ee06fdcf6b/test/metabase/query_processor_test/date_time_zone_functions_test.clj#L701

File: modules/drivers/mongo/src/metabase/driver/mongo.clj (hunk 1 above)
Comment (calherries) [reply modules/drivers/mongo/src/metabase/driver/mongo.clj]: Thanks, updated here https://github.com/metabase/metabase/pull/27042/commits/7175935adf224dc9410b6e059fae5ad8dde82492

File: modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj [hunk 2]
```diff
@@ -549,6 +549,44 @@
                     :unit      unit
                     :amount    amount}})
 
+(defmethod ->rvalue :datetime-diff [[_ x y unit]]
+  (do
```
Comment (metamben) [inline modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:553]: Why is this needed?

File: modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj [hunk 3]
```diff
@@ -549,6 +549,44 @@
...
+          y (->rvalue y)]
+      (case unit
+        :year
+        {$divide [(->rvalue [:datetime-diff x y :month]) 12]}
+
+        :quarter
+        {$divide [(->rvalue [:datetime-diff x y :month]) 3]}
+
+        :week
+        {$divide [(->rvalue [:datetime-diff x y :day]) 7]}
```
Comment (metamben) [inline modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj:565]: We do a little extra work calling `->rvalue` twice on `x` and `y`, I wonder if it makes sense to separate the recursive cases from the terminal ones.

File: modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj (hunk 2 above)
Comment (calherries) [reply modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj]: It's not, thanks. I've removed it

File: modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj (hunk 3 above)
Comment (calherries) [reply modules/drivers/mongo/src/metabase/driver/mongo/query_processor.clj]: True. Updated https://github.com/metabase/metabase/pull/27042/commits/51f1529d991279a0c3d437d13a2c11933de0396d

**Review by camsaul — APPROVED** [review]
I think this would be a little easier to read if you broke that datetime-diff helper out into a multimethod with separate methods for the different units like we have in the SQL QP. Otherwise this LGTM

**Comment by calherries** [comment]
@camsaul FYI I created a datetime-diff helper multimethod as you suggested and merged
